balrog/api.go

```
@@ -13,220 +13,50 @@ package main
-  // parse the signature into R and S value by splitting it in the middle
-  sig.R, sig.S = new(big.Int), new(big.Int)
-  sig.R.SetBytes(data[:len(data)/2])
-  sig.S.SetBytes(data[len(data)/2:])
-  sig.Finished = true
-  return sig
+	if uintptr(loggerFunc) == 0 {
+		return 0, errors.New("no logger initialized")
+	}
+	message := C.CString(string(p))
+	C.callLogger(loggerFunc, l.level, message)
+	C.free(unsafe.Pointer(message))
+	return len(p), nil
 }
 
 //export balrogSetLogger
 func balrogSetLogger(loggerFn uintptr) {
-  loggerFunc = unsafe.Pointer(loggerFn)
-}
-
-//export balrogValidateSignature
-func balrogValidateSignature(publicKey string, signature string, input string) bool {
-  logger := log.New(&CLogger{level: 0}, "", 0)
-
-  cs := Unmarshal(signature)
-  if cs == nil {
-    logger.Println("Unmarshal failed")
-    return false
-  }
-
-  block, _ := pem.Decode([]byte(publicKey))
-  if block == nil {
-    logger.Println("Invalid PEM public key format")
-    return false
-  }
-
-  keyInterface, err := x509.ParsePKIXPublicKey(block.Bytes)
-  if err != nil {
-    logger.Println("Failed to parse public key DIR:", err)
-    return false
-  }
-  pubkey := keyInterface.(*ecdsa.PublicKey)
-
-  // verify signature on input data
-  if !cs.VerifyData([]byte(input), pubkey) {
-    logger.Println("Failed to verify content signature")
-    return false
-  }
-
-  logger.Println("Verified")
-  return true
+	loggerFunc = unsafe.Pointer(loggerFn)
+}
+
+//export balrogValidate
+func balrogValidate(x5u string, input string, signature string, rootHash string, leafCertSubject string) bool {
+	logger := log.New(&CLogger{level: 0}, "", 0)
+	err := contentsignature.Verify([]byte(input), []byte(x5u), signature, rootHash)
+	if err != nil {
+		logger.Println("Verification failed with error:", err.Error())
+		return false
+	}
+	// Verify method does not verify the leaf cert subject, so we do it here.
+	certs, _ := contentsignature.ParseChain([]byte(x5u))
```

ignoring the error because `contentsignature.Verify` already calls ParseChain.